src/daft-json/src/local.rs

```diff
@@ -0,0 +1,485 @@
+    pub fn finish(&self) -> DaftResult<Table> {
+        let mut bytes = self.bytes;
+        let mut n_threads = self.n_threads;
+        let mut total_rows = 128;
+
+        if let Some((mean, std)) = get_line_stats_json(bytes, self.sample_size) {
+            let line_length_upper_bound = mean + 1.1 * std;
+
+            total_rows = (bytes.len() as f32 / (mean - 0.01 * std)) as usize;
+            if let Some(n_rows) = self.n_rows {
+                total_rows = std::cmp::min(n_rows, total_rows);
+                // the guessed upper bound of the no. of bytes in the file
+                let n_bytes = (line_length_upper_bound * (n_rows as f32)) as usize;
+
+                if n_bytes < bytes.len() {
+                    if let Some(pos) = next_line_position(&bytes[n_bytes..]) {
+                        bytes = &bytes[..n_bytes + pos]
+                    }
+                }
+            }
+        }
+
+        if total_rows <= 128 {
+            n_threads = 1;
+        }
+
+        let total_len = bytes.len();
+        let chunk_size = self.chunk_size.unwrap_or_else(|| total_len / n_threads);
+        let file_chunks = self.get_file_chunks(bytes, n_threads, total_len, chunk_size);
+
+        let tbls = self.pool.install(|| {
+            file_chunks
+                .into_par_iter()
+                .map(|(start, stop)| {
+                    let chunk = &bytes[start..stop];
+                    self.parse_json_chunk(chunk, chunk_size)
+                })
+                .collect::<DaftResult<Vec<Table>>>()
+        })?;
+
+        let tbl = tables_concat(tbls)?;
+
+        // The `limit` is not guaranteed, so we need to properly apply the limit after concatenating the
+        if let Some(limit) = self.n_rows {
+            if tbl.len() > limit {
+                return tbl.head(limit);
+            }
+        };
+        Ok(tbl)
+    }
+
+    fn parse_json_chunk(&self, bytes: &[u8], chunk_size: usize) -> DaftResult<Table> {
+        let mut scratch = vec![];
+        let scratch = &mut scratch;
+
+        let daft_fields = Arc::new(
+            self.schema
+                .fields
+                .values()
+                .map(|f| Arc::new(f.clone()))
+                .collect::<Vec<_>>(),
+        );
+        let arrow_schema = self.schema.to_arrow()?;
+
+        // The `RawValue` is a pointer to the original JSON string and does not perform any deserialization.
+        // This is a trick to use the line-based deserializer from serde_json to iterate over the lines
+        // This is more accurate than using a `Lines` iterator.
+        // Ideally, we would instead use a line-based deserializer from simd_json, but that is not available.
+        let iter =
+            serde_json::Deserializer::from_slice(bytes).into_iter::<&serde_json::value::RawValue>();
+
+        let mut columns = arrow_schema
+            .fields
+            .iter()
+            .map(|f| {
+                (
+                    Cow::Owned(f.name.to_string()),
```

Comment:
I don't _think_ this needs to be a `Cow` since we should always have an owned `String` when building with this field iterator, and you can query an `IndexMap<String, _>` with a `&str` key, if needed.

Reply:
So this is because the indexmap isnt `IndexMap<String, _>`, but `IndexMap<Cow<'_, str>, _>`

https://github.com/universalmind303/Daft/blob/ea395ddd59d37aee024e8cfdbf18bdca755a673a/src/daft-json/src/deserializer.rs#L6

I wanted to avoid changing that type.